### ISSUE TITLE
Zeitwerk is not compatible with Rails Inflector

### ISSUE THREAD

**Issue by deepj** [fxn/zeitwerk#80]
I have some defined acronyms in `config/inflections.rb` which are used in namespaces. Typically something like `Id` => `ID`

If I have the following concern `app/models/concerns/my_acronym/test.rb` where I have module `MyACRONYM::Test`. Then I include this concern in `app/models/application_record.rb` like `include MyACRONYM::Test`. With the classic autoloader it works but not with Zeitwerk. I'm getting `uninitialized constant MyACRONYM::Test (NameError)`.

This happens after Rails 5.2 => 6.0 upgrade.

**Comment by fxn** [comment]
By default, the inflector used in Rails is Active Support `camelize`, and acronyms are supported. If you have defined:

```ruby
ActiveSupport::Inflector.inflections(:en) do |inflect|
  inflect.acronym 'MyACRONYM'
end
```

Then classic infers the file name from the constant to be autoloaded, using `underscore`:

```ruby
"MyACRONYM".underscore # => "myacronym"
```

(Please, note that underscoring `MyACRONYM` yields `myacronym`, not `my_acronym`.)

If autoloading with Zeitwerk, you go the other way around: from file name to constant name:

```ruby
"myacronym".camelize # => "MyACRONYM"
```

Can please create an absolutely minimal application that has your exact acronym defined in the initializers and one file that is not autoloaded as expected?

**Comment by deepj** [comment]
@fxn 

Nope, there is no `MyACRONYM` defined in the Rails inflector. I only defined ACRONYM in the inflector.

```
ActiveSupport::Inflector.inflections(:en) do |inflect|
  inflect.acronym 'ACRONYM'
end
```

So I combine a regular word and upper-cased ACRONYM. In this case it can be ID instead of Id`.

An example can be CloudID namespace. So the directory is `cloud_id`. Then I can have a module `CloudID::Person` which is located in `app/models/concerns/cloud_id/person.rb`. In this case Zeitwerk doesn't find the module.


I'll try to prepare a minimal app but probably tomorrow.

**Comment by fxn** [comment]
I believe there has to be something more going on in your app.

Rails [sets a custom inflector](https://github.com/rails/rails/blob/8ab4fd12f18203b83d0f252db96d10731485ff6a/railties/lib/rails/autoloaders.rb#L10-L26), which is one [based on `String#camelize`](https://github.com/rails/rails/blob/65f9112d9601b5b6342e021a4a03353d15f7b5ef/activesupport/lib/active_support/dependencies/zeitwerk_integration.rb#L56-L60). With this simple trick, inflection handling is automatically inherited by Zeitwerk.

In particular, the support for acronyms within `String#camelize` just works.

Areproduction of the situation described above (as I understand it) autoloads  fine:

```
$ cat config/initializers/inflections.rb
ActiveSupport::Inflector.inflections(:en) do |inflect|
  inflect.acronym 'ID'
end

$ cat app/models/application_record.rb
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true
  include CloudID::Person
end

$ cat app/models/concerns/cloud_id/person.rb
module CloudID::Person; end

$ bin/rails runner 'p ApplicationRecord'
ApplicationRecord(abstract)

$ bin/rails runner 'p CloudID::Person'
CloudID::Person
```

**Comment by deepj** [comment]
@fxn it seems you're right. I created a new Rails 6.0 app and I can't see the problem. Is there any Rails' 5.x settings having affect on Zeitwerk? We have this in `config/application.rb`

```
    config.autoload_paths += Dir["#{config.root}/lib"]
    config.eager_load_paths += Dir["#{config.root}/lib"]
```

**Comment by fxn** [comment]
Let's see if we understand it. Can you throw

```
Rails.autoloaders.log!
```

in _config/application.rb_ after the line that sets the defaults, trigger the error, and share the output?

**Comment by fxn** [comment]
What does `”cloud_id”.camelize` return in your app?

**Comment by fxn** [comment]
Hey!

I like to keep issues and PRs to 0 in this project. Since the conversation stopped, and I doubt Zeitwerk itself is not working properly, I am going to close it.

If you'd like to investigate this further in the context of your Rails application, please do open an issue in the Rails repo and /cc me.

**Comment by deepj** [comment]
@fxn yes, I couldn't have answered since meanwhile I got inflammation and fever and I have to use antibiotics. There are serious reason why someone can't answer immediately. Sad :(

Anyway, you're right. The problem in the app. But I discovered anything else. Rails inflection has effects on loaded gems. And I guess it can cause huge problems if inflections are applied to them.

Easy reproduction with Rails 6.0

1. `rails new new_app`
2. modify `config/initializers/inflections.rb` and add `ID` acronym
2. modify `config/environments/development.rb` and set `config.eager_load` to `true`
3. `rails s`

Then the issue is occurred:

```
~/.gem/ruby/2.5.3/gems/actionmailbox-6.0.0/app/models/action_mailbox/inbound_email.rb:30:in `<class:InboundEmail>': uninitialized constant #<Class:0x00007fdcac82b938>::MessageId (NameError)
Did you mean?  ActionMailbox::InboundEmail::MessageID
```

**Comment by fxn** [comment]
Hey! Don't get me wrong, it is totally fine to be silent for an indefinite period of time.

I meant it in the other direction: Was continuing the thread here as a courtesy because it was opened here and had velocity, albeit it did not belong to this project. Since velocity halted, there was no reason to continue here, the Rails repo is the right one.

I'll send you how to configure a custom inflector later.

**Comment by fxn** [comment]
I have documented [how to define custom inflectors](https://edgeguides.rubyonrails.org/autoloading_and_reloading_constants.html#customizing-inflections) in a Rails application.

The example is trivial, its purpose is to show the basic technique. Of course, inflectors can use lookup tables instead, or be as sophisticated as needed.

Please excuse the misundestarding, and I hope you get well.

**Comment by deepj** [comment]
@fxn Thanks. I'm a bit better now.

Thanks for a tip on the custom inflector for Zeitwerk. Though, I still see this as an issue and now I'm not sure where to report it. I guess into Rails' Issues. In our case, the most acronyms are not for classes/namespaces purposes. And it's unpredictable when we'll add a new gem if it breaks something. My thoughts are here if it's possible to isolate Zeitwerk's applying Rails' inflections only for app paths by default in Rails.

**Comment by fxn** [comment]
You're experiencing the brittleness of Active Support inflections. They are global, so if you change anything, there is the potential issue of affecting engines, or gems using Active Support.

See, only in the framework:

```
rails (master)$ ag --ruby camelize | wc -l
     152
```

That is why inflection rules have been frozen for many years.

For example, if you publish an engine as a gem, you cannot be 100% sure the inflections configured in the parent application will break your engine in some way. However, parent applications with conflicting rules won't work with the engine, so hopefully that will be detected soon. For example, if the engine implements `HtmlController` an acronym may break its routes.

Classic autoloading uses `underscore`, which is subject to this gotcha too. Except, in `classic` you cannot change the inflector!

Bottom line: What Rails doesn't support is inflection rules scoped by app or gem. AS inflections are global and that won't change. Autoloading in `zeitwerk` mode has `String#camelize` as _default_ inflector, but also gives you full control and the third of the three examples in the guide depicts that. You have even the absolute path in the second argument.

**Comment by fxn** [comment]
Let me add that Zeitwerk itself does not use Active Support, it ships with a [deterministic inflector](https://github.com/fxn/zeitwerk/blob/master/lib/zeitwerk/inflector.rb) that can't be affected by any external configuration. Individual loaders have their own independent inflector, so they have full control about how they load their projects.

In particular, gems using Zeitwerk won't be affected by changes in the inflection rules (unless of course the author configured `String#camelize` by hand).

Now, _the Rails integration_ that implements autoloading based on Zeitwerk [sets a custom inflector](https://github.com/rails/rails/blob/a1b6c1669fc99e20439d6a8cada36392c82c48a8/railties/lib/rails/autoloaders.rb#L10-L26) by default because the natural default for Rails programmers is `String#camelize`.

Engines don't have their own inflectors, the parent application is the one in charge of autoloading.
